compiler/passes/normalize.cpp

```diff
@@ -2683,6 +2745,70 @@ static bool skipFixup(ArgSymbol* formal, Expr* domExpr, Expr* eltExpr) {
   return false;
 }
 
+static bool isArrayFormal(ArgSymbol* arg) {
+  if (BlockStmt* typeExpr = arg->typeExpr) {
+    //
+    // The body is usually a single callExpr.  However there are rare
+    // cases in which normalization generates one or more call_temps
+    // i.e. a sequence of defExpr/primMove pairs.
+    //
+    // In either case the desired callExpr is the tail of the body.
+    //
+
+    if (CallExpr* call = toCallExpr(typeExpr->body.tail)) {
+      if (call->isNamed("chpl__buildArrayRuntimeType")) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+static void fixupExportedArrayFormals(FnSymbol* fn) {
+  CallExpr* retCall = toCallExpr(fn->body->body.tail);
+  INT_ASSERT(retCall);
+  CallExpr* callOrigFn = retCall;
+  if (retCall->isPrimitive(PRIM_RETURN)) {
+    INT_ASSERT(retCall->numActuals() == 1);
+    callOrigFn = toCallExpr(retCall->get(1));
+  }
+
+  INT_ASSERT(callOrigFn);
+
+  for_formals(formal, fn) {
+    if (isArrayFormal(formal)) {
+      BlockStmt*            typeExpr = formal->typeExpr;
+
+      CallExpr*             call     = toCallExpr(typeExpr->body.tail);
+      int                   nArgs    = call->numActuals();
+      Expr*                 eltExpr  = nArgs == 2 ? call->get(2) : NULL;
+
+      if (eltExpr == NULL) {
+        USR_FATAL(formal, "array argument '%s' in exported function '%s'"
+                  " must specify its type", formal->name, fn->name);
+        continue;
+      }
+
+      // Create a representation of the array argument that is accessible
+      // outside of Chapel (in the form of a pointer and a corresponding size)
+      formal->typeExpr->replace(new BlockStmt(new CallExpr("_type_construct_chpl_external_array")));
```

Comment:
This is OK for now if it works but AFAIK a reasonable alternative would be to make chpl_external_array a "well known type" and then put here a SymExpr referring to its TypeSymbol. That seems "cleaner" to be but I understand such things are subjective. But, if it seems better to you too, it's probably worth doing.